urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/client/GrinderThread.java

```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ */
+package grinder.client;
+
+import grinder.client.user.BrowserType;
+import grinder.client.user.User;
+
+import java.sql.Timestamp;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.UUID;
+
+/**
+ * Thread executing the actual grinding.
+ */
+public class GrinderThread extends Thread {
+
+	Client client;
+	private boolean testRunning;
+
+	/**
+	 * Instantiates a new grinder thread.
+	 *
+	 * @param client reference
+	 */
+	public GrinderThread(final Client c, final String name) {
+
+		super(name);
+		client = c;
+
+		testRunning = false;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.lang.Thread#run()
+	 */
+	@Override
+	public void run() {
+
+		// #1 BLOCKING until start-time is reached
+		while (!testRunning)
+		{
+			// in case somebody panic'ed cleanup and stop immediately
+			if ((client.getState() == ClientState.CLEANUP))
+			{
+				client.cleanupUsers();
+				client.setState(ClientState.IDLE);
+			}
+
+			// if time to start is reached
+			if ((client.getState() == ClientState.ARMED)
+					&& client.testTicket.getStartTime().before(
+							new Timestamp(System.currentTimeMillis())))
+			{
+				client.setState(ClientState.TESTING);
+				client.cgui.logMessage("BEGINNING TESTING!");
+				testRunning = true;
+			}
+		}
+		
+		client.setAllUsersInitialized(false);
+
+		// #2 GRINDING
+		while (client.getState() == ClientState.TESTING)
+		{
+
+			// generate/restock user threads
+			try
+			{
+				while (client.userHTTP.size() < client.testTicket.getUserCountHTTP())
+				{
+					client.addUser(BrowserType.HTTP);
+					Thread.sleep(10);
+					if (client.getState() == ClientState.CLEANUP) break;
+				}
+				while (client.userWS.size() < client.testTicket.getUserCountWebSocket())
+				{
+					client.addUser(BrowserType.WEBSOCKET);
+					Thread.sleep(10);
+					if (client.getState() == ClientState.CLEANUP) break;
+				}
+				while (client.userHTTP.size() < client.testTicket.getUserCountStratus())
+				{
+					client.addUser(BrowserType.STRATUS);
+					Thread.sleep(10);
+					if (client.getState() == ClientState.CLEANUP) break;
+				}
+			} catch (final InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)